### PR TITLE
[Bugfix] fix OpenAI API server startup with --disable-frontend-multiprocessing

### PR DESCRIPTION
**Pull request by dtrifiro**
#8491 this causes the http server to always error out when `--disable-frontend-multiprocessing` is enabled (for example when using embedding models):

```
ERROR:    [Errno 98] error while attempting to bind on address ('0.0.0.0', 8000): address already in use
```

By passing the socket fd to uvicorn, we solve the problem.

### CODE REVIEW

**Comment by kevin314** [comment]
Taking a look now to see if I can recreate the issue on my end

**Comment by kevin314** [comment]
@dtrifiro Did you use any args for starting the api server other than specifying the port

**Comment by youkaichao** [comment]
@dtrifiro what's your environment? this fix passed all our ci tests. I don't know why it does not work in your environment.

**Comment by nickandbro** [comment]
> This seemed strange to me aswell, but by further investigating, I discovered this only happens with embedding models:
> 
> ```
> vllm serve intfloat/e5-mistral-7b-instruct
> ```
> 
> I'll investigate some more and try to come up with a fix.

This has also been happening on my end as well with that particular embedding model

**Comment by dtrifiro** [comment]
Update: this happens whenever frontend multiprocessing is disabled (e.g. for embedding models) `--disable-frontend-multiprocessing`

File: tests/entrypoints/openai/test_basic.py [hunk 1]
```diff
@@ -35,6 +73,14 @@ async def client(server):
         yield async_client
 
 
+@pytest.mark.parametrize(
```
Comment (dtrifiro) [inline tests/entrypoints/openai/test_basic.py:76]: This should run `test_show_version` twice, once with the default args and once with `--disable-frontend-multiprocessing`, to make sure that the server actually starts in that case.

File: vllm/entrypoints/openai/api_server.py [hunk 2]
```diff
@@ -547,6 +548,7 @@ def signal_handler(*_) -> None:
             ssl_certfile=args.ssl_certfile,
             ssl_ca_certs=args.ssl_ca_certs,
             ssl_cert_reqs=args.ssl_cert_reqs,
+            fd=sock.fileno(),
```
Comment (youkaichao) [inline vllm/entrypoints/openai/api_server.py:551]: wow this fix is pretty cool!

**Review by youkaichao — APPROVED** [review]

File: vllm/entrypoints/openai/api_server.py [hunk 3]
```diff
@@ -520,8 +520,11 @@ async def run_server(args, **uvicorn_kwargs) -> None:
     logger.info("vLLM API server version %s", VLLM_VERSION)
     logger.info("args: %s", args)
 
-    temp_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-    temp_socket.bind(("", args.port))
```
Comment (youkaichao) [inline vllm/entrypoints/openai/api_server.py:524]: out of curiosity, I still don't understand why this does not work when frontend multiprocessing is disabled.

Comment (dtrifiro) [reply vllm/entrypoints/openai/api_server.py:524]: @youkaichao 

Creating the engine causes the executor to be instantiated and initialized in a new process using `multiprocessing`.

Doing this after the socket is created and bound, causes a reference to the socket being passed onto the child, causing resources not to be freed properly.

Here's a minimal example that reproduces the issue:


```python
import asyncio
import uvloop
import socket
import uvicorn
import multiprocessing
import time

from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()

mp_method = "fork"
# mp_method = "spawn"
mp = multiprocessing.get_context(mp_method)


class Request(BaseModel):
    text: str


@app.post("/")
async def echo(request: Request):
    return {"request": request.text}


def worker(name: str, *args):
    """dummy worker"""
    time.sleep(5)


def spawn_process(*args):
    process = mp.Process(
        target=worker,
        name="worker",
        args=("process_worker", *args),
        # kwargs={}
        daemon=True,
    )
    process.start()
    return process


HOST, PORT = "0.0.0.0", 8000


async def serve_http():
    config = uvicorn.Config(
        app=app,
        host=HOST,
        port=PORT,
        timeout_keep_alive=5,
    )
    server = uvicorn.Server(config)

    loop = asyncio.get_running_loop()
    server_task = loop.create_task(server.serve())

    await server_task


async def run_server():
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

    s.bind((HOST, PORT))

    process = spawn_process() # in vllm, this block is `build_async_engine_client`
    print(f"spawned {process=}")

    s.close()

    await serve_http()

    process.join()


if __name__ == "__main__":
    uvloop.run(run_server())

```

**Comment by dtrifiro** [comment]
Anything holding this back? @youkaichao

**Comment by russellb** [comment]
FYI, it seems this PR introduced some new unexpected behavior, see #9737

**Comment by russellb** [comment]
> FYI, it seems this PR introduced some new unexpected behavior, see #9737

PR #10012 has a follow-up to this change.

**Pull request merged**